### PR TITLE
Move code related to zenith pageserver to contrib/zenith:

### PR DESCRIPTION
**Pull request by lubennikovaav**
1. Move  code related to zenith pageserver communication to contrib/zenith.
To use it add 'shared_preload_library = zenith' to postgresql.conf;

2. Move zenith specific GUCs to custom GUC variables defined in zenith library.
Now they require 'zenith' prefix: i.e  zenith.page_server_connstring = 'host=127.0.0.1 port=64000'

Add custom GUC zenith.wal_redo  (type boolean) to indicate postgres running in wal-redo mode.

3. Add smgr_hook that can be used to define custom smgrs.
Remove smgrsw[] array and smgr_sw selector. Instead, smgropen() uploads f_smgr implementation using smgr_hook.
Also add smgr_init_hook and smgr_shutdown_hook. And a lot of mechanical changes in smgr.c functions.

4.  Remove zenith_clog code. Most likely we won't use it, so it isn't worth porting to the new API.

### CODE REVIEW

File: contrib/zenith/libpagestore.c [hunk 1]
```diff
@@ -178,15 +179,72 @@ zenith_call(ZenithRequest request)
 	return (ZenithResponse *) resp;
 }
 
+
+// static bool
```
Comment (hlinnaka) [inline contrib/zenith/libpagestore.c:183]: would be nice to keep this working

**Review by hlinnaka — APPROVED** [review]
LGTM. More work remains but definitely a step in the right direction.

File: contrib/zenith/libpagestore.c (hunk 1 above)
Comment (lubennikovaav) [reply contrib/zenith/libpagestore.c]: fixed

File: contrib/zenith/pagestore_smgr.c [hunk 2]
```diff
@@ -646,7 +610,7 @@ zenith_read(SMgrRelation reln, ForkNumber forkNum, BlockNumber blkno,
 					forkNum)));
 	}
 	memcpy(buffer, resp->page, BLCKSZ);
-	((PageHeader)buffer)->pd_flags &= ~PD_WAL_LOGGED; /* Clear PD_WAL_LOGGED bit stored in WAL record */
+	// ((PageHeader)buffer)->pd_flags &= ~PD_WAL_LOGGED; /* Clear PD_WAL_LOGGED bit stored in WAL record */
```
Comment (hlinnaka) [inline contrib/zenith/pagestore_smgr.c:613]: Why this change?

File: src/backend/tcop/zenith_wal_redo.c [hunk 3]
```diff
@@ -206,6 +205,9 @@ WalRedoMain(int argc, char *argv[],
 	/* read control file (error checking and contains config ) */
 	LocalProcessControlFile(false);
 
+	/* upload zenith shared library */
```
Comment (hlinnaka) [inline src/backend/tcop/zenith_wal_redo.c:208]: "upload" doesn't sound right

File: contrib/zenith/pagestore_smgr.c [hunk 4]
```diff
@@ -236,9 +238,9 @@ zenith_load(void)
 {
 	Assert(page_server_connstring && page_server_connstring[0]);
 
-	load_file("libpqpagestore", false);
-	if (page_server == NULL)
-		elog(ERROR, "libpqpagestore didn't initialize correctly");
+	// load_file("libpqpagestore", false);
```
Comment (hlinnaka) [inline contrib/zenith/pagestore_smgr.c:241]: I guess this is not needed becuse this is now a contrib module that can use libpq functions directly. So I think this can be removed competely, not just commented out

File: contrib/zenith/pagestore_smgr.c [hunk 5]
```diff
@@ -252,74 +254,36 @@ zenith_wallog_page(SMgrRelation reln, ForkNumber forknum, BlockNumber blocknum,
 		zenith_load();
 
 	/*
-	 * If the page was not WAL-logged before eviction then we can lose its modification.
+	 * If the page was not WAL-logged before eviction then we can loose this modification.
```
Comment (hlinnaka) [inline contrib/zenith/pagestore_smgr.c:257]: "lose" is correct

**Review by hlinnaka — APPROVED** [review]

File: contrib/zenith/libpagestore.c [hunk 6]
```diff
@@ -178,15 +181,75 @@ zenith_call(ZenithRequest request)
...
+							 NULL, NULL, NULL);
+
 	if (page_server != NULL)
 		zenith_log(ERROR, "libpqpagestore already loaded");
 
 	zenith_log(PqPageStoreTrace, "libpqpagestore already loaded");
 	page_server = &api;
+
+	//Is there more correct way to pass CustomGUC to postgres code?
+	zenith_timeline_walproposer = zenith_timeline;
```
Comment (knizhnik) [inline contrib/zenith/libpagestore.c:241]: Why do we need zenith_timeline variable at all? Why not to use zenith_timeline_walproposer directly in DefineCustomStringVariable?
